flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SsgNetworkMemoryCalculationUtils.java

```diff
@@ -148,6 +161,78 @@ private static TaskInputsOutputsDescriptor buildTaskInputsOutputsDescriptor(
         return ret;
     }
 
+    private static Map<IntermediateDataSetID, Integer> getMaxInputChannelNumsForDynamicGraph(
+            ExecutionJobVertex ejv) {
+
+        Map<IntermediateDataSetID, Integer> ret = new HashMap<>();
+        for (IntermediateResult consumedResult : ejv.getInputs()) {
+            ret.put(consumedResult.getId(), getMaxInputChannelNumForResult(ejv, consumedResult));
+        }
+
+        return ret;
+    }
+
+    private static Map<IntermediateDataSetID, Integer> getMaxSubpartitionNumsForDynamicGraph(
+            ExecutionJobVertex ejv) {
+
+        Map<IntermediateDataSetID, Integer> ret = new HashMap<>();
+
+        for (IntermediateResult intermediateResult : ejv.getProducedDataSets()) {
+            final int maxNum =
+                    Arrays.stream(intermediateResult.getPartitions())
+                            .map(IntermediateResultPartition::getNumberOfSubpartitions)
+                            .reduce(0, Integer::max);
+            ret.put(intermediateResult.getId(), maxNum);
+        }
+
+        return ret;
+    }
+
+    @VisibleForTesting
+    static int getMaxInputChannelNumForResult(
+            ExecutionJobVertex ejv, IntermediateResult consumedResult) {
+        DistributionPattern distributionPattern = consumedResult.getConsumingDistributionPattern();
+
+        if (distributionPattern == DistributionPattern.ALL_TO_ALL) {
+            int numChannelsToConsumePerPartition =
+                    getMaxNumOfChannelsForConsuming(consumedResult.getPartitions()[0]);
+            int numConsumedPartitions = consumedResult.getNumberOfAssignedPartitions();
+            return numChannelsToConsumePerPartition * numConsumedPartitions;
+
+        } else if (distributionPattern == DistributionPattern.POINTWISE) {
+            int numPartitions = consumedResult.getNumberOfAssignedPartitions();
+            int numConsumers = ejv.getParallelism();
+            // when using dynamic graph, all partitions have the same number of subpartitions
```

Comment:
When enable dynamic graph, the num of subpartitions will be compute by `IntermediateResultPartition#computeNumberOfMaxPossiblePartitionConsumers`.  Whether the consumer parallelism is set or not, the returned value of`computeNumberOfMaxPossiblePartitionConsumers` should be the same for different partitions.

Reply:
I took a look at IntermediateResultPartition again and yes you are right. But I just feel that this assumption is easy to break and easy to lost track. This is because it is more like an implicit implementation result, rather than an explicit contract. 

e.g. if later someone tried to optimized `IntermediateResultPartition#computeNumberOfSubpartitions()` to avoid creating extra subpartitions if the parallelism of the downstream vertex is known already, this assumption may break. However, he/she may not be aware of the changes here the assumption here may stay unchanged.

Reply:
Another thing I'm thinking is that, can we reuse the logic of `getMaxInputChannelNums`? Because the edge connections should be the same no matter if it's a dynamic graph or not. The only difference is the subpartition range that one edge consumes. Furthermore, execution edges from the same `IntermediateResult` to the same sub-task consumes the same range.

Reply:
You are right. I will change it to use `TaskDeploymentDescriptorFactory#computeConsumedSubpartitionRange` to compute subpartition range for each consumer, and then calculate the max number of channels according to the subpartition range.

Reply:
> e.g. if later someone tried to optimized `IntermediateResultPartition#computeNumberOfSubpartitions()` to avoid creating extra subpartitions if the parallelism of the downstream vertex is known already, this assumption may break. However, he/she may not be aware of the changes here the assumption here may stay unchanged.

You are right. I will change it to use TaskDeploymentDescriptorFactory#computeConsumedSubpartitionRange to compute subpartition range for each consumer, and then calculate the max number of channels according to the subpartition range.